tensorboard/components/vz_line_chart2/log-scale.ts

```diff
@@ -0,0 +1,164 @@
+limitations under the License.
+==============================================================================*/
+namespace vz_line_chart2 {
+
+const MIN_VALUE = 1e-15;
+
+function log(x: number): number {
+  return Math.log(x) / Math.log(10);
+}
+
+function pow(x: number): number {
+  return Math.pow(10, x);
+}
+
+export class LogScale extends TfScale {
+  private _d3LogScale = d3.scaleLog();
+  private _untransformedDomain: number[];
+
+  constructor() {
+    super();
+    this.padProportion(.2);
+  }
+
+  public scale(x: number): number {
+    // Returning NaN makes sure line plot does not plot illegal values.
+    if (x <= 0) return NaN;
+    return this._d3LogScale(x);
+  }
+
+  public invert(x: number): number {
+    return this._d3LogScale.invert(x);
+  }
+
+  public scaleTransformation(value: number) {
+    return this.scale(value);
+  }
+
+  public invertedTransformation(value: number) {
+    return this.invert(value);
+  }
+
+  public getTransformationDomain(): [number, number] {
+    return this.domain() as [number, number];
+  }
+
+  protected _getDomain() {
+    return this._untransformedDomain;
+  }
+
+  protected _setDomain(values: number[]) {
+    this._untransformedDomain = values;
+    const [min, max] = values;
+    super._setDomain([Math.max(MIN_VALUE, min), max]);
+  }
+
+  /**
+   * Given a domain, pad it and clip the lower bound to MIN_VALUE.
+   */
+  protected _niceDomain(domain: number[], count?: number): number[] {
+    const [low, high] = domain;
+    const adjustedLogLow = Math.max(log(MIN_VALUE), log(low));
+    const logHigh = log(high);
+    const pad = (logHigh - adjustedLogLow) * this.padProportion();
+    const logLowFloor = Math.floor(adjustedLogLow);
+    const logHighCeil = Math.ceil(logHigh);
+    return [
+      pow(Math.max(log(MIN_VALUE), adjustedLogLow - pad, logLowFloor)),
+      pow(Math.min(logHigh + pad, logHighCeil)),
+    ];
+  }
+
+  /**
+   * Generates a possible extent based on data from all plots the scale is
+   * connected to by taking the minimum and maximum values of all extents for
+   * lower and upper bound, respectively.
+   * @override to remove default padding logic.
+   */
+  protected _getUnboundedExtent(ignoreAttachState): number[] {
+    const includedValues = this._getAllIncludedValues(ignoreAttachState);
+    let extent = this._defaultExtent();
+    if (includedValues.length !== 0) {
+      const combinedExtent = [
+        Plottable.Utils.Math.min<number>(includedValues, extent[0]),
+        Plottable.Utils.Math.max<number>(includedValues, extent[1]),
+      ];
+      extent = this._niceDomain(combinedExtent);
+    }
+    return extent;
+  }
+
+  protected _getAllIncludedValues(ignoreAttachState = false): number[] {
+    const values = super._getAllIncludedValues();
+    // For log scale, the value cannot be smaller or equal to 0. They are
+    // negative infinity.
+    return values.map(x => x > 0 ? x : MIN_VALUE);
```

Comment:
Hmm this method is not returning values for the plot. It is returning datasets values of all plots for sake of calculating the bounds. Since this log scale cannot show numbers smaller than some arbitrary `x` we use, it is correct that we clip these values here.